/* Copyright (C) 2014-2015 Alexander Shishenko <alex@shishenko.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
syntax = "proto3";
package librevault.protocol.v2;

enum MessageType {
  HANDSHAKE = 0;
  CHOKE = 1;
  UNCHOKE = 2;
  INTERESTED = 3;
  NOTINTERESTED = 4;
  INDEXUPDATE = 5;
  METAREQUEST = 6;
  METARESPONSE = 7;
  BLOCKREQUEST = 8;
  BLOCKRESPONSE = 9;
}

message ProtocolMessage {
  MessageType type = 1;
  bytes payload = 2;
}

/* Message payloads */
message Handshake {
  bytes auth_token = 1;
  string device_name = 2;
  string user_agent = 3;
  uint32 dht_port = 4;
}

message IndexUpdate {
  bytes path_id = 1;
  uint64 revision = 2;
  bytes bitfield = 3;
}

message MetaRequest {
  bytes path_id = 1;
  uint64 revision = 2;
}

message MetaResponse {
  bytes meta = 1;
  bytes signature = 2;
  bytes bitfield = 3;
}

message BlockRequest {
  bytes ct_hash = 1;
  uint64 offset = 2;
  uint32 length = 3;
}

message BlockResponse {
  bytes ct_hash = 1;
  uint64 offset = 2;
  bytes content = 3;
}
